### PR TITLE
[Access] Upgrade lru cache

### PR DESCRIPTION
**Pull request by nozim**
Following up with #4565, upgrade all the dependencies on lru.Cache with generic version for enhanced performance.

### CODE REVIEW

File: engine/access/rpc/connection/connection_test.go [hunk 1]
```diff
@@ -614,8 +612,7 @@ func TestExecutionEvictingCacheClients(t *testing.T) {
 	ctx := context.Background()
 
 	// Retrieve the cached client from the cache
-	result, _ := cache.Get(clientAddress)
-	cachedClient := result.(*CachedClient)
+	cachedClient, _ := cache.Get(clientAddress)
```
Comment (peterargue) [inline engine/access/rpc/connection/connection_test.go:615]:
```suggestion
	cachedClient, ok := cache.Get(clientAddress)
	require.True(t, ok)
```

File: engine/access/rpc/connection/connection_test.go [hunk 2]
```diff
@@ -671,7 +668,7 @@ func TestCircuitBreakerExecutionNode(t *testing.T) {
 
 	// Set the connection pool cache size.
 	cacheSize := 1
-	connectionCache, _ := lru.New(cacheSize)
+	connectionCache, _ := lru.New[string, *CachedClient](cacheSize)
```
Comment (peterargue) [inline engine/access/rpc/connection/connection_test.go:671]:
```suggestion
	connectionCache, err := lru.New[string, *CachedClient](cacheSize)
	require.NoError(t, err)
```

File: engine/access/rpc/connection/connection_test.go [hunk 3]
```diff
@@ -755,8 +752,8 @@ func TestCircuitBreakerCollectionNode(t *testing.T) {
 
 	// Set the connection pool cache size.
 	cacheSize := 1
-	connectionCache, _ := lru.New(cacheSize)
 
+	connectionCache, _ := lru.New[string, *CachedClient](cacheSize)
```
Comment (peterargue) [inline engine/access/rpc/connection/connection_test.go:756]:
```suggestion
	connectionCache, err := lru.New[string, *CachedClient](cacheSize)
	require.NoError(t, err)
```

**Review by peterargue — COMMENTED** [review]
thanks @nozim! This will clean things up nicely

File: engine/access/rpc/connection/connection_test.go [hunk 4]
```diff
@@ -4,29 +4,24 @@ import (
...
 	"testing"
 	"time"
 
-	"github.com/sony/gobreaker"
-
-	"go.uber.org/atomic"
-	"pgregory.net/rapid"
-
-	lru "github.com/hashicorp/golang-lru"
+	lru "github.com/hashicorp/golang-lru/v2"
```
[the remark this answers is not in the document]
Comment (nozim) [reply engine/access/rpc/connection/connection_test.go:11]: sorry, I'm struggling with very stubborn Goland goimports settings. But will enforce this manually!

File: fvm/storage/derived/derived_chain_data.go [hunk 5]
```diff
@@ -4,8 +4,7 @@ import (
 	"fmt"
 	"sync"
 
-	"github.com/hashicorp/golang-lru/simplelru"
-
+	"github.com/hashicorp/golang-lru/v2/simplelru"
 	"github.com/onflow/flow-go/model/flow"
```
Comment (sideninja) [inline fvm/storage/derived/derived_chain_data.go:8]:
```suggestion

	"github.com/onflow/flow-go/model/flow"
```

File: engine/access/rpc/backend/backend.go [hunk 6]
```diff
@@ -2,12 +2,14 @@ package backend
 
 import (
 	"context"
+	"crypto/md5"
```
Comment (sideninja) [inline engine/access/rpc/backend/backend.go:5]:
```suggestion
	"crypto/md5" // #nosec
```

**Comment by nozim** [comment]
@sideninja thank you, fixes on their way :)

File: engine/access/rpc/backend/backend.go (hunk 6 above)
Comment (nozim) [reply engine/access/rpc/backend/backend.go]: Fixed here https://github.com/onflow/flow-go/pull/4605/commits/7a2cbb4a8b52d2c386d59405b5d5fcc74d7ab40e

File: fvm/storage/derived/derived_chain_data.go (hunk 5 above)
Comment (nozim) [reply fvm/storage/derived/derived_chain_data.go:9]: fixed here https://github.com/onflow/flow-go/pull/4605/commits/fc26728abdd52cdfa7cc37695729127bddc544f6

File: engine/access/rpc/backend/backend.go [hunk 7]
```diff
@@ -119,7 +118,7 @@ func New(
 		retry.Activate()
 	}
 
-	loggedScripts, err := lru.New(DefaultLoggedScriptsCacheSize)
+	loggedScripts, err := lru.New[[16]byte, time.Time](DefaultLoggedScriptsCacheSize)
```
[the remark this answers is not in the document]
Comment (nozim) [reply engine/access/rpc/backend/backend.go]: https://github.com/onflow/flow-go/pull/4605/commits/4ef02f3dda67af38ce89c4b0bc639c2fc9482afb

File: engine/access/rpc/backend/backend_scripts.go [hunk 8]
```diff
@@ -27,13 +27,14 @@ import (
...
-	log                  zerolog.Logger
-	metrics              module.BackendScriptsMetrics
-	loggedScripts        *lru.Cache
+	headers           storage.Headers
+	executionReceipts storage.ExecutionReceipts
+	state             protocol.State
+	connFactory       connection.ConnectionFactory
+	log               zerolog.Logger
+	metrics           module.BackendScriptsMetrics
+	loggedScripts     *lru.Cache[[16]byte, time.Time]
```
[the remark this answers is not in the document]
Comment (nozim) [reply engine/access/rpc/backend/backend_scripts.go]: https://github.com/onflow/flow-go/pull/4605/commits/4ef02f3dda67af38ce89c4b0bc639c2fc9482afb

File: engine/access/rpc/connection/connection_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nozim) [reply engine/access/rpc/connection/connection_test.go:11]: https://github.com/onflow/flow-go/pull/4605/commits/659e22f12550fccc8b37697ea98b97e457be91d2

File: engine/access/rpc/connection/connection_test.go [hunk 9]
```diff
@@ -129,8 +124,8 @@ func TestProxyAccessAPIConnectionReuse(t *testing.T) {
 	connectionFactory.CollectionGRPCPort = cn.port
 	// set the connection pool cache size
 	cacheSize := 1
-	cache, _ := lru.NewWithEvict(cacheSize, func(_, evictedValue interface{}) {
-		evictedValue.(*CachedClient).Close()
+	cache, _ := lru.NewWithEvict[string, *CachedClient](cacheSize, func(_ string, client *CachedClient) {
```
[the remark this answers is not in the document]
Comment (nozim) [reply engine/access/rpc/connection/connection_test.go]: https://github.com/onflow/flow-go/pull/4605/commits/e1695557fa595606badd277611b21325f29b999c

File: engine/access/rpc/connection/connection_test.go (hunk 2 above)
Comment (nozim) [reply engine/access/rpc/connection/connection_test.go]: https://github.com/onflow/flow-go/pull/4605/commits/72d13303f587af13b3ea3e1d1e7c9432d61d9514

File: engine/access/rpc/connection/connection_test.go (hunk 3 above)
Comment (nozim) [reply engine/access/rpc/connection/connection_test.go]: https://github.com/onflow/flow-go/pull/4605/commits/72d13303f587af13b3ea3e1d1e7c9432d61d9514

File: engine/access/rpc/connection/connection_test.go (hunk 1 above)
Comment (nozim) [reply engine/access/rpc/connection/connection_test.go]: https://github.com/onflow/flow-go/pull/4605/commits/72d13303f587af13b3ea3e1d1e7c9432d61d9514

**Review by sideninja — APPROVED** [review]

**Review by jordanschalm — APPROVED** [review]
Very nice cleanup, thank you!

**Comment by jordanschalm** [comment]
@nozim The CI is failing because the generated mocks do not exactly match the committed mocks (the imports are not in the same order). 

If you run `make generate-mocks` from the repository root, it should fix the mock files.

File: engine/access/rpc/connection/connection_test.go [hunk 10]
```diff
@@ -671,8 +656,8 @@ func TestCircuitBreakerExecutionNode(t *testing.T) {
 
 	// Set the connection pool cache size.
 	cacheSize := 1
-	connectionCache, _ := lru.New(cacheSize)
-
+	connectionCache, err := lru.New[string, *CachedClient](cacheSize)
+	require.Nil(t, err)
```
[the remark this answers is not in the document]
Comment (nozim) [reply engine/access/rpc/connection/connection_test.go]: yup, amended here https://github.com/onflow/flow-go/pull/4605/commits/8ff9fb849f377be864233e47b29432cbda24967b. Thanks

File: engine/access/rpc/connection/connection_test.go [hunk 11]
```diff
@@ -755,8 +740,9 @@ func TestCircuitBreakerCollectionNode(t *testing.T) {
 
 	// Set the connection pool cache size.
 	cacheSize := 1
-	connectionCache, _ := lru.New(cacheSize)
 
+	connectionCache, err := lru.New[string, *CachedClient](cacheSize)
+	require.Nil(t, err)
```
[the remark this answers is not in the document]
Comment (nozim) [reply engine/access/rpc/connection/connection_test.go]: same https://github.com/onflow/flow-go/pull/4605/commits/8ff9fb849f377be864233e47b29432cbda24967b ;)

**Comment by jordanschalm** [comment]
Hey @nozim, since the last review the diff has grown quite a bit. The additions don't seem related this PR. Was there an inadvertent merge of some other branch? It seems like the extra unrelated changes were added in this commit: https://github.com/onflow/flow-go/pull/4605/commits/c2938e46b5ee8c6fad25404130a00e3dd135e5d5.

**Comment by nozim** [comment]
@jordanschalm yes, I'll revert those changes after the other one is merged. Thanks for bringing this up 🙏

**Comment by nozim** [comment]
@jordanschalm removed irrelevant changes. Kindly re-review.

**Comment by nozim** [comment]
@jordanschalm kindly help to rerun the ci 🙏

**Review by peterargue — APPROVED** [review]

File: engine/access/rpc/backend/backend.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (nozim) [reply engine/access/rpc/backend/backend.go]: https://github.com/onflow/flow-go/pull/4605/commits/9a6979879cce833a39ddd48008abafb41eced3eb

**Comment by nozim** [comment]
@peterargue kindly run the ci. Fixed the annotation

**Comment by nozim** [comment]
Recreated this PR on latest master here https://github.com/onflow/flow-go/pull/4700. So closing this one.

**Pull request closed without merging**